docs/TemplateFiles.md

```diff
@@ -0,0 +1,129 @@
+
+Larger template files may seem intimidating at first. This document is to help describe the basics of template files so you can start crafting your own.
+
+We're reference a few different types of stream classes here but we won't go into too much detail about how they work in this document. Please refer to the Beginner's Guide for some basics about different stream types.
+
+## Your First Template File
+
+Let's start with a very basic template file.
+
+```
+_type: Tumblr\StreamBuilder\Streams\NullStream
+```
+
+That's it! It doesn't do much, but that's an entire template file which uses the built-in StreamBuilder `NullStream` class (which never returns any stream elements). You could load this template into an application and it would 100% function.
+
+In this case, all we've done is define the type (with the `_type:` property) of the stream. The type of a stream will always be a PHP class. It may be one of the built-in classes or it may be a custom class you've written yourself.
+
+## Expanding Our Template
+
+YAML files handle hierarchy through indentation. That means, as we build on our template file, we will keep indenting our primary stream so it gets pushed further to the right.
+
+Let's update our initial template to have a filter as an example:
+
+```
+_type: Tumblr\StreamBuilder\Streams\FilteredStream
+stream:
+	_type: Tumblr\StreamBuilder\Streams\NullStream
+stream_filter:
+	_type: MyApplication\StreamFilters\MyFilter
+```
+
+Notice how we've essentially wrapped our original `NullStream` so it's _inside_ the `FilteredStream`. 
+
+The primary `_type` property of our file is now the `FilteredStream` and our `NullStream` is inside the filter under the `stream:` property. The `FilteredStream` also has an additional `stream_filter` property which defines the class we're using to handle the actual filtering.
+
+**NOTE: It's important to see that any template file property that's defining a stream class must have a `_type` property.**
+
+## Stream Options
+
+Let's digress for just a moment to talk about template file properties that are available for different stream types.
+
+Most stream classes support varying properties, some required, some optional. Most of the required properties are things like `_type`, `stream`, `inner`, etc.
+
+At this point, the easiest way to see the options available is to look at the class definition for the stream type.
+
+Let's look at the [`FilteredStream`](https://github.com/Automattic/stream-builder/blob/main/lib/Tumblr/StreamBuilder/Streams/FilteredStream.php#L86) as an example:
+
+```
+public function __construct(
+        Stream $inner,
+        StreamFilter $filter,
+        string $identity,
+        int $retry_count = null,
+        float $overfetch_ratio = null,
+        bool $skip_filters = false,
+        bool $slice_result = true
+    ) {
+```
```

Comment:
can you link to the "from_template" method instead? `from_template` will give you the name of the properties